networkit/cpp/generators/ClusteredRandomGraphGenerator.cpp

```diff
@@ -3,55 +3,128 @@
+
+    if (!isProbability(pIntra))
+        throw std::runtime_error("Error: pIntra must be in [0, 1]");
+
+    if (!isProbability(pInter))
+        throw std::runtime_error("Error: pInter must be in [0, 1]");
+
+    if (pIntra < pInter)
+        WARN("Intra-cluster probability is lower than inter-cluster probability.");
+
+    zeta.setUpperBound(k);
 }
 
 Graph ClusteredRandomGraphGenerator::generate() {
-    assert(pin >= pout);
+    GraphBuilder graphBuilder(n);
 
-    Graph G(n);
-    // assign nodes evenly to clusters
-    Partition zeta(n);
-    zeta.setUpperBound(k);
-    G.forNodes([&](node v){
-        index c = Aux::Random::integer(k-1);
-        zeta.addToSubset(c, v);
-    });
-
-    assert (zeta.numberOfSubsets() == k);
-
-    G.forNodePairs([&](node u, node v){
-        if (zeta.subsetOf(u) == zeta.subsetOf(v)) {
-            if (Aux::Random::probability() <= pin) {
-                G.addEdge(u, v);
-            }
-        } else {
-            if (Aux::Random::probability() <= pout) {
-                G.addEdge(u, v);
-            }
+    // Sequence of all nodes grouped by cluster. Here the vertices of each cluster are stored
+    // sequentially one after the other
+    std::vector<node> clustersSequence;
+
+    // Index in `clustersSequence` where each cluster ends. E.g., clusterEnd[i] is the index in
+    // `clustersSequence` of the first vertex in the (i + 1)-th cluster.
+    std::vector<size_t> clusterEnd(k);
+
+    {
+        // To assign vertices to clusters uniformly at random
+        std::uniform_int_distribution<index> clusterDistr{0, k - 1};
+
+        // Nodes grouped by cluster
+        std::vector<std::vector<node>> clusters(k);
+
+        std::vector<omp_lock_t> locks(k);
+        for (index i = 0; i < k; ++i)
+            omp_init_lock(&locks[i]);
+
+        // Assign nodes to clusters
+        graphBuilder.parallelForNodes([&zeta = zeta, &clusters, &clusterDistr, &locks](node u) {
+            const index cluster = clusterDistr(Aux::Random::getURNG());
+            zeta.addToSubset(cluster, u);
+            omp_set_lock(&locks[cluster]);
+            clusters[cluster].push_back(u);
+            omp_unset_lock(&locks[cluster]);
+        });
+
+        // Last cluster ends at n, each cluster i ends at
+        // n - \sum_{j = k - i}^{k} cluster[j].size()
+        clusterEnd[k - 1] = n;
+        for (int64_t i = k - 2; i >= 0; --i)
+            clusterEnd[i] = clusterEnd[i + 1] - clusters[i + 1].size();
+
+        // From a 2-D vector to a single 1-D vector.
+        for (auto &&cluster : clusters)
+            clustersSequence.insert(clustersSequence.end(), cluster.begin(), cluster.end());
+    }
+
+    // Add all the half-out edges from a node u to all the other nodes in the given sequence.
+    // To avoid adding multiple half-out edges, with respect to `clustersSequence` all the nodes
+    // in [start, end) should be subsequent to `u`.
```

Comment:
This description contradicts the implementation, the implementation never adds an edge to the node at position `start`. In the usage below, this correctly avoids the self-loop to `u`, as below - in contradiction to the documentation -, `start` is not subsequent to `u`, but then it passes the end of the cluster (i.e., the next node after the cluster) as `start`, and thus the edge from `u` to the next node after the cluster is impossible as far as I can see.